pkg/log/syslog/syslog.go

```diff
@@ -0,0 +1,268 @@
+// limitations under the License.
+
+package syslog
+
+import (
+	"io"
+	"net"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/RackSec/srslog"
+	"github.com/Sirupsen/logrus"
+)
+
+type SyslogConfig struct {
+	Network   string
+	RAddr     string
+	Tag       string
+	Priority  srslog.Priority
+	Formatter Formatter
+}
+
+type dialer interface {
+	dial(*SyslogConfig) (Writer, error)
+}
+
+type Hook struct {
+	entries chan *logrus.Entry
+	writer  Writer
+	cfg     SyslogConfig
+
+	// for testing
+	running chan struct{}
+}
+
+type Writer interface {
+	io.WriteCloser
+
+	Emerg(string) error
+	Crit(string) error
+	Err(string) error
+	Warning(string) error
+	Info(string) error
+	Debug(string) error
+}
+
+const maxLogBuffer = 100
+
+type Formatter int
+
+const (
+	RFC3164 Formatter = iota
+)
+
+func NewHook(cfg *SyslogConfig, d dialer) (*Hook, error) {
+	hook := &Hook{
+		entries: make(chan *logrus.Entry, maxLogBuffer),
+		running: make(chan struct{}),
+	}
+
+	if d == nil {
+		d = &defaultDialer{
+			d: &syslogDialer{},
+		}
+	}
+
+	var err error
+	hook.writer, err = d.dial(cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	return hook, nil
+}
+
+type defaultDialer struct {
+	// for testing purposes
+	d dialer
+}
+
+func (d *defaultDialer) dial(cfg *SyslogConfig) (Writer, error) {
+	w, err := d.d.dial(cfg)
+	if err != nil {
+		switch err.(type) {
+		case *net.AddrError, *net.ParseError:
+			return nil, err
+		}
+	}
+
+	return &writerWrapper{
+		writer: w,
+		dialer: d.d,
+		cfg:    *cfg,
+	}, nil
+}
+
+type syslogDialer struct{}
+
+func (d *syslogDialer) dial(cfg *SyslogConfig) (Writer, error) {
+	w, err := srslog.Dial(cfg.Network, cfg.RAddr, cfg.Priority, cfg.Tag)
+	if w != nil {
+		switch cfg.Formatter {
+		case RFC3164:
+			w.SetFormatter(srslog.RFC3164Formatter)
+		}
+
+		return w, nil
+	}
+
+	return nil, err
+}
+
+func (hook *Hook) Fire(entry *logrus.Entry) error {
+	select {
+	case hook.entries <- entry:
+	default:
+		// drop log entry
+	}
+
+	return nil
+}
+
+func (hook *Hook) Levels() []logrus.Level {
+	return logrus.AllLevels
+}
+
+func (hook *Hook) Run() {
+	close(hook.running)
+	for entry := range hook.entries {
+		hook.writeEntry(entry)
+	}
+
+	logrus.Warnf("exited syslog loop")
+}
+
+func (hook *Hook) writeEntry(entry *logrus.Entry) error {
+	// just use the message since the timestamp
+	// is added by the syslog package
+	line := entry.Message
+
+	switch entry.Level {
+	case logrus.PanicLevel:
+		return hook.writer.Crit(line)
+	case logrus.FatalLevel:
+		return hook.writer.Crit(line)
+	case logrus.ErrorLevel:
+		return hook.writer.Err(line)
+	case logrus.WarnLevel:
+		return hook.writer.Warning(line)
+	case logrus.InfoLevel:
+		return hook.writer.Info(line)
+	case logrus.DebugLevel:
+		return hook.writer.Debug(line)
+	}
+
+	return nil
+}
+
+type writerWrapper struct {
+	writer Writer
+	cfg    SyslogConfig
+	dialer dialer
+}
+
+func (w *writerWrapper) withRetry(f func()) {
+	if w.writer == nil {
+		w.writer, _ = w.dialer.dial(&w.cfg)
```

Comment:
Actually I should correct that. logrus *does not* grab a lock before calling the hooks. It does, however, grab a lock when calling `Write` on the `io.Writer` it uses. The hook implementation is ok since the `Fire` calls just write to a channel. `writerWrapper` is however not thread-safe. This is OK for now, since it is only called by the hook. For future uses we can document that this is not thread-safe, and have any caller protect calls made to it.